### PR TITLE
Fix dependency object types returned by custom dependency detection

### LINKED ISSUE #2905: dependency('gl') can return an object which claims to be a pkgconfig dependency, but isn't
```meson.build
project('gl dependency', 'c')

gl_dep = dependency('gl', method: 'pkg-config', required: false)
message(gl_dep.type_name())
if gl_dep.found() and gl_dep.type_name() == 'pkgconfig'
  gl_dep.get_pkgconfig_variable('prefix')
endif
```

confusingly fails with:

```
Message: pkgconfig

Meson encountered an error in file meson.build, line 6, column 0:
'gl' is not a pkgconfig dependency
```

### PR DESCRIPTION
**Pull request by jon-turney**
Return a PkgConfgDependency object from custom dependency detection which uses pkgconfig. This allows get_pkgconfig_variable() to work correctly on the result.

Return a ConfigToolDependency subclass object from custom dependency detection which uses a config tool.  This allows get_configtool_variable() to work correctly on the result.

Closes #2905

### CODE REVIEW

File: mesonbuild/dependencies/misc.py [hunk 1]
```diff
@@ -810,90 +811,80 @@ def get_pkgconfig_variable(self, variable_name, kwargs):
...
         else:
             return [DependencyMethods.PKGCONFIG, DependencyMethods.CONFIG_TOOL]
 
-    def get_pcap_lib_version(self):
-        return self.compiler.get_return_value('pcap_lib_version', 'string',
-                                              '#include <pcap.h>', self.env, [], [self])
+    @staticmethod
+    def get_pcap_lib_version(ctdep):
+        return ctdep.compiler.get_return_value('pcap_lib_version', 'string',
+                                              '#include <pcap.h>', ctdep.env, [], [ctdep])
```
Comment (ignatenkobrain) [inline mesonbuild/dependencies/misc.py]: __[Flake8]__

[E128] continuation line under-indented for visual indent

**Comment by jon-turney** [comment]
@dcbaker I've probably misunderstood something about how ConfigToolDependency pickling is supposed to be working, so if you can clarify, that would be a great help.

**Comment by dcbaker** [comment]
@jon-turney. There's some some ordering issues with pickling them, basically if the pickle is de-serialized before the ConfigToolDependency class is loaded. I was looking at this at one point (I was working on the same thing you are), but I got side tracked. I think the problem is that the `__reduce__` method of ConfigToolDependency is a closure, and if it wasn't that things would work okay. I can poke at it a bit mroe tomorrow if you don't get it figured out before then.

**Comment by jon-turney** [comment]
> @jon-turney. There's some some ordering issues with pickling them, basically if the pickle is de-serialized before the ConfigToolDependency class is loaded things break. I was looking at this at one point (I was working on the same thing you are), but I got side tracked. I think the problem is that the __reduce__ method of ConfigToolDependency is a closure, and if it wasn't that things would work okay. I can poke at it a bit mroe tomorrow if you don't get it figured out before then.

Thanks for the explanation.  Well, I think I have something that unpickles these correctly, your opinion may vary :)

**Comment by dcbaker** [comment]
That's a lot less code than what I had, and I don't see any obvious corners in the approach. `pickle` is rather finiky, and meta programming is hard. The two together are painful.

**Comment by dcbaker** [comment]
I'm not an official reviewer, but I rather like this series, I wrote something similar at one point, and I think this is both a nice cleanup and removes a lot of unexpected behavior.

**Comment by nirbheek** [comment]
I think this should go in for the 0.45 release.

**Pull request merged**